xds/internal/xdsclient/clientimpl.go

```diff
@@ -69,22 +145,57 @@ func (c *clientImpl) BootstrapConfig() *bootstrap.Config {
 	return c.config
 }
 
-// close closes the gRPC connection to the management server.
+// close closes the xDS client and releases all resources.
 func (c *clientImpl) close() {
 	if c.done.HasFired() {
 		return
 	}
 	c.done.Fire()
-	// TODO: Should we invoke the registered callbacks here with an error that
-	// the client is closed?
 
-	c.authorityMu.Lock()
+	c.topLevelAuthority.close()
 	for _, a := range c.authorities {
 		a.close()
 	}
-	c.idleAuthorities.Clear(true)
-	c.authorityMu.Unlock()
+
+	// Channel close cannot be invoked with the lock held, because it can race
+	// with stream failure happening at the same time. The latter will callback
+	// into the clientImpl and will attempt to grab the lock. This will result
+	// in a deadlock. So instead, we release the lock and wait for all active
+	// channel to be closed.
+	var channelsToClose []*xdsChannel
+	c.channelsMu.Lock()
+	for _, cs := range c.xdsActiveChannels {
+		channelsToClose = append(channelsToClose, cs.channel)
+	}
+	c.xdsActiveChannels = nil
+	c.channelsMu.Unlock()
+
+	var wg sync.WaitGroup
+	for _, c := range channelsToClose {
+		wg.Add(1)
+		go func(c *xdsChannel) {
+			defer wg.Done()
+			c.close()
+		}(c)
```

Comment:
What does this spawning of the goroutine buy you here?

You already gave up lock so this is equivalent to just doing it inline and waiting for that to process.

Is the main reason you can close in parallel and it's faster? But if it calls back it'll be in serial anyway. I'm fine with it just wondering what the point is.

Reply:
I had some reason to spawn the goroutine. It's been a long time since I rewrote this part of the code. But now, I don't see why. So, I'm getting rid of it.

Reply:
I'm fine with getting rid of the goroutine. Is there some style guide here? It feels more complex but allows the work to be done in parallel. I'm assuming we want to keep goroutines as lean as possible, even though they're pretty cheap?

Reply:
> Is there some style guide here?

Not that I know of.

> even though they're pretty cheap?

Yes, given that they are pretty cheap, the general recommendation is to spawn them whenever and wherever required. But if the same thing can be done without a goroutine and it is simpler, then I would do it without the goroutine. And as I mentioned in another comment, this is happening during xDS client close. So, we shouldn't be worrying about performance here, but should focus on more readable and maintainable code here.

Reply:
Yeah makes sense thanks.